cpp/include/cudf/column/column.hpp

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/types.hpp>
+#include "column_view.hpp"
+
+#include <rmm/device_buffer.hpp>
+#include <rmm/mr/device_memory_resource.hpp>
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+
+class column {
+ public:
+  ~column() = default;
+  column& operator=(column const& other) = delete;
+  column& operator=(column&& other) = delete;
+  column() = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new column by deep copying the device memory of another
+   * column.
+   *
+   * All device memory allocation and copying is done using the
+   * `device_memory_resource` and `stream` from `other`.
+   *
+   * @param other The other column to copy
+   *---------------------------------------------------------------------------**/
+  column(column const& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new column object from the contents of another `column`
+   * using move semantics. After the move, `other` is guaranteed to be empty.
+   *
+   * @param other The other column whose device memory will be moved to the new
+   * column
+   *---------------------------------------------------------------------------**/
+  column(column&& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new column
+   *
+   * @param[in] dtype The element type
+   * @param[in] size The number of elements in the column
+   * @param[in] data `device_buffer` whose contents will be copied for the
+   * column's data
+   * @param[in] null_mask Optional, `device_buffer` whose contents will be
+   * copied for the column's null mask. Buffer may be empty if `null_count` is 0
+   * or `UNKNOWN_NULL_COUNT`.
+   * @param null_count Optional, the count of null elements. If unknown, specify
+   * `UNKNOWN_NULL_COUNT` to indicate that the null count should be computed on
+   * the first invocation of `null_count()`.
+   * @param children Optional, vector of child columns
+   *---------------------------------------------------------------------------**/
+  template <typename B1, typename B2 = rmm::device_buffer>
```

Comment:
It's to allow for the device buffers to be "universal references", i.e., you can pass an l-value reference (deep copy) or an r-value reference (move) into the same function template. Otherwise, I'd need 4 distinct APIs allowing for all combinations of passing the data and null mask via l-value or r-value reference.